prov/verbs/src/fi_verbs.h

```diff
@@ -487,11 +487,11 @@ struct fi_ibv_ep {
 	struct fi_ibv_srq_ep	*srq_ep;
 	struct fi_info		*info;
 	/* TODO: it would be removed */
-	struct fi_ibv_dgram_buf_pool	grh_pool;
+	struct fi_ibv_dgram_buf_pool	*grh_pool;
 
-	struct ibv_send_wr	rma_wr;
-	struct ibv_send_wr	msg_wr;
-	struct ibv_sge		sge;
+	struct ibv_send_wr	*rma_wr;
+	struct ibv_send_wr	*msg_wr;
+	struct ibv_sge		*sge;
```

Comment:
If all of these are allocated/freed as a group, you could create a new structure with the fields embedded in that, and use a single pointer to that structure.  This would simplify the creation/destruction process, plus reduce the ep structure even further.  It looks like you can do this for the work request related structs -- not sure about the grh_pool.

Reply:
Good catch. Thanks!

I'm going to remove grh_pool soon. I think verbs/DGRAM can require to support `FI_MSG_PREFIX` in application code. So, it should improve code and avoid double allocation from different pools (RxD - from its TX pool buffer and Verbs/DGRAM - from grh_pool). RxD will have to allocate space for message prefix (needed size to keep GRH buffer)

Reply:
Do we see an improvement in latency numbers after reducing the size of ep?

Reply:
This wouldn't be a latency improvement, but a memory consumption/scalability enhancement.  It adds a pointer deference into the code path, which would actually increase the latency, though the instruction count would likely end up being the same.

Reply:
Got it. This is per msg_ep. Makes sense. Thanks.

Reply:
> This is per msg_ep

Yes, this is main goal.

> though the instruction count would likely end up being the same

You're right, I don't see increasing of instruction count.